### DISCUSSION TITLE
Can wasmFS exist as a general read-only, desktop, linkable virtual FS?

### DISCUSSION THREAD

**Discussion by vadimkantorov** [emscripten-core/emscripten#20764] [General]
E.g. I need a read-only, virtual FS to read from ISO or a ZIP or an area in the executable's binary file (e.g. packaged via `ld` or `#embed`). I wonder, @sbc100  can wasmFS be used as such an impl in a regular desktop libc context (e.g. with muslc)?

Basically, it would need to redirect to memopen/fmemopen all file ops that start from some specially-prefixed path and either override open/fopen functions and have a script to patch/ar custom static muslc into such redirects.

This seems similar to some functionality of wasmFS, so curious if it can be used for such a purpose?

**Comment by kripken** [comment]
By "regular desktop libc context" do you mean outside of wasm and emscripten? Just built normally?

If so then with some adaptation it could, perhaps. It does implement in-memory files. But it currently implements syscalls from the custom musl libc used in Emscripten, which would not work out of the box with another libc. You'd need to add support for syscalls for whatever other libc you want (and some other minor things like removing usage of `emscripten_*` APIs, but there is little of that).

**Reply by vadimkantorov** [reply]
> Just built normally?

Yeah!

This is interesting, but I guess this would require quite a lot of work, I might try at some point... Does wasmFS support some read-only mode with all files available as a pointer + number_of_bytes or as mmap'd pointer? (e.g. if we're mmapping a portion of the binary executable itself)

I guess for this mode only a limited number of syscalls is needed: open/openat/stat and maybe some others.

If factoring it out of emscripten as extremely compact virtual FS code is possible, it might find usage/development in some other contexts as well!

**Reply by kripken** [reply]
mmap support is quite limited atm, so I'm not sure if that would work. But supporting more of mmap would be easier in a read-only mode, that's true (no need to worry about the file growing and need to realloc the storage), but I don't think we have such a read-only backend atm (adding one might not be hard though).

**Reply by vadimkantorov** [reply]
E.g. a frequent usage mode is providing a ZIP file (mmapable, but in compressed modes would require some decompression) / TAR File (very mmapable - even individual files are mmapable) / ISO (same) file as the read-only FS contents. And such read-only FS portable C-lib is useful for all kinds of portable apps/distribution scenario. Also embedding these binary files directly in an executable (at least ELF allows this) can be useful when file size is not a problem.



E.g. in a similar scenario, there exists Cosmopolitan libc's [`zipos`](https://github.com/jart/cosmopolitan/blob/master/libc/runtime/zipos-open.c), but it also seems tied to this  custom libc.

E.g. like this, a portable single-file python can be built similar to https://ahgamut.github.io/2021/07/13/ape-python/ by packing all module files in a tar/iso and embedding it into the binary, but with using regular muslc instead of APE.

**Reply by vadimkantorov** [reply]
Within the emscripten world, more regualr ISO/TAR/maybe others formats can then replace file_packager in some scenarios (read-only is also somewhat frequent scenario for the pre-packed files)

**Reply by sbc100** [reply]
If one wanted to preload some other format (other than the file_packger format) it should be easy enough to do.  The file_packager format is not actually something that is intrinsic to emscripten, its just some `--pre-js` code that the `file_pakager.py` script generates.  If one wanted to load a `.iso` instead that should also be possible with a `--pre-js` script.

**Reply by vadimkantorov** [reply]
Thanks for a suggestion! Though, ideally, I'd like to use the same virtual FS lib both for wasm target and for x64/static-musl target. When the usecase is a single-file, portable app, then using the same lib would be very useful
